File: interviews/ForUsAll/Python/setBitPermutation.py
class SetBits:

    BitsSetTable = [0 for _ in range(256)] 

    for i in range(256):
        BitsSetTable[i] = (i & 1) + BitsSetTable[i // 2] 

    @classmethod
    def lookUp(self, num):

        return SetBits.BitsSetTable[num & 0xFF] +\
            SetBits.BitsSetTable[(num >> 8) & 0xFF] +\
            SetBits.BitsSetTable[(num >> 16) & 0xFF] +\
            SetBits.BitsSetTable[(num >> 24) & 0xFF] 

class Solution:

    def findAllSetBitPermutationNum(self, num):

        setBitNum = SetBits.lookUp(num) 

        return self.setBitPermutation(setBitNum) 

    def setBitPermutation(self, setBitNum):

        results = [] 

        if setBitNum < 0 or setBitNum > 32:
            return results 

        self.dfs(setBitNum, 0, [], results) 

        return results
        

    def dfs(self, setBitNum, index, permutation, results):

        if setBitNum == 0:

            permutation = permutation[::-1] 
            results.append(self.bitsToNum(permutation))
            return 

        if index == 32 or 32 - index < setBitNum:
            return 

        permutation.append(0) 
        self.dfs(setBitNum, index + 1, permutation, results) 
        permutation.pop() 

        permutation.append(1)  
        self.dfs(setBitNum - 1, index + 1, permutation, results) 
        permutation.pop() 

    def bitsToNum(self, bits):

        num = 0 

        for bit in bits:

            num = num * 2 + bit 

        return num 

solution = Solution() 

print(solution.findAllSetBitPermutationNum(8))    

File: interviews/FUA/Java/setBitPermutation.java
import java.util.ArrayList;

public class setBitsPermutation{

    public static void main(String[] args) {

        Solution solution = new Solution(); 

        System.out.println(solution.findSetBitPermutationNumbers(1));
        
    }
}

class SetBitTable {

    private int[] table = new int[256]; 

    public SetBitTable() {

        for (int index = 0; index < 256; index++) {

            this.table[index] = index & 1 + table[index / 2]; 
        }
    
    }
    
    
    public int lookup(int num) {

        return this.table[num & 0xFF] + this.table[(num >> 8) & 0xFF] + this.table[(num >> 16) & 0xFF] + this.table[(num >> 24) & 0xFF]; 
    }
}

class Solution {

    public ArrayList<Integer> findSetBitPermutationNumbers(int num) {
        SetBitTable table = new SetBitTable();
        int setBitsNum =  table.lookup(num); 
        return this.setBitsPermutation(setBitsNum); 

    }

    private ArrayList<Integer> setBitsPermutation(int setBitsNum) {

        ArrayList<Integer> results = new ArrayList<>(); 
        ArrayList<Integer> permutation = new ArrayList<>(); 

        this.dfs(setBitsNum, 0, permutation, results); 
        
        return results; 
    }

    private void dfs(int setBitsNum, 
                     int index, 
                     ArrayList<Integer> permutation, 
                     ArrayList<Integer> results) {

            if (setBitsNum == 0) {

                results.add(this.bitsToNum(permutation)); 
                return; 
            }

            if (index == 31 || 31 - index < setBitsNum) {
                return; 
            }

            permutation.add(1); 
            this.dfs(setBitsNum - 1, index + 1, permutation, results); 
            permutation.remove(permutation.size() - 1); 

            permutation.add(0); 
            this.dfs(setBitsNum, index + 1, permutation, results); 
            permutation.remove(permutation.size() - 1); 

    }

    private int bitsToNum(ArrayList<Integer> bits) {

        int num = 0; 

        for (int index = bits.size() - 1; index >= 0; index--){

            num = num * 2 + bits.get(index);  
        }

        return num; 

    }

}

